### PR TITLE
Move assumeUnique helper function to ocean.core.TypeConvert

### PR DESCRIPTION
**Pull request by joseph-wakeling-sociomantic**
This helper function is not merely transitional: it will continue to be useful in D2-only applications.

A selective `public import` has been added to `ocean.transition` so as to avoid any breaking change for downstream apps and libraries, but the change should allow projects moving to D2 only to have a migration path away from `ocean.transition` while continuing to use `assumeUnique`.

### CODE REVIEW

File: src/ocean/core/TypeConvert.d [hunk 1]
```diff
@@ -22,6 +22,50 @@ import ocean.meta.traits.Indirections;
...
+    input = null;
+    return cast(Immut!(T)[]) tmp;
+}
+
+unittest
+{
+    auto s1 = "aaa".dup;
+    auto s2 = assumeUnique(s1);
+    test!("==")(s2, "aaa");
+    test!("is")(s1, null);
```
Comment (mihails-strasuns) [inline src/ocean/core/TypeConvert.d]: `null` -> `mstring.init` to fix D1 build

**Comment by joseph-wakeling-sociomantic** [comment]
@mihails-strasuns fixed.  I've also replaced the `ocean.transition` import in `ocean.core.TypeConvert` to avoid any possibility of circular dependencies:

```diff
diff --git a/src/ocean/core/TypeConvert.d b/src/ocean/core/TypeConvert.d
index 046d0cfe..96070f56 100644
--- a/src/ocean/core/TypeConvert.d
+++ b/src/ocean/core/TypeConvert.d
@@ -17,7 +17,8 @@ module ocean.core.TypeConvert;
 
 
 
-import ocean.transition;
+import ocean.meta.types.Qualifiers;
+import ocean.meta.types.Typedef;
 import ocean.meta.traits.Indirections;
 
 version ( UnitTest ) import ocean.core.Test;
@@ -62,7 +63,7 @@ unittest
     auto s1 = "aaa".dup;
     auto s2 = assumeUnique(s1);
     test!("==")(s2, "aaa");
-    test!("is")(s1, null);
+    test!("is")(s1, mstring.init);
 }
```

**Comment by joseph-wakeling-sociomantic** [comment]
The codecov report is a dirty rotten liar.  We have exactly the same tests!

**Comment by joseph-wakeling-sociomantic** [comment]
Tweaked slightly to retain alphabetical order of imports:

```diff
diff --git a/src/ocean/core/TypeConvert.d b/src/ocean/core/TypeConvert.d
index 96070f56..ddc87c03 100644
--- a/src/ocean/core/TypeConvert.d
+++ b/src/ocean/core/TypeConvert.d
@@ -17,9 +17,9 @@ module ocean.core.TypeConvert;
 
 
 
+import ocean.meta.traits.Indirections;
 import ocean.meta.types.Qualifiers;
 import ocean.meta.types.Typedef;
-import ocean.meta.traits.Indirections;
 
 version ( UnitTest ) import ocean.core.Test;
```

**Comment by joseph-wakeling-sociomantic** [comment]
Just as a minor implementation note: my first draft tried to use `d1to2fix_inject` to make this `auto ref` for D2 code (on the grounds that we might as well make it easy to support lvalues for D2-only apps/libs even if we can't do it for D1 code).

That failed because `d1to2fix_inject` does not support injecting `auto` (I assume for good and deliberate reason), but I tested the validity of `auto ref` by manually editing the D2 code, and things worked.  So this would conceivably be a way to bring the function's behaviour closer to Phobos if there is willingness to support it.

**Pull request merged**